File: src/Swift.Bindings/tests/IntegrationTests/FunctionalTests/Generics/GenericTests.cs
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

using Swift.GenericTests;
using Swift.Runtime;
using Xunit;

namespace BindingsGeneration.FunctionalTests
{
    public class GenericsTests : IClassFixture<GenericsTests.TestFixture>
    {
        private readonly TestFixture _fixture;

        public GenericsTests(TestFixture fixture)
        {
            _fixture = fixture;
        }

        public class TestFixture
        {
            static TestFixture()
            {
                InitializeResources();
            }

            private static void InitializeResources()
            {
                // Initialize
            }
        }

        [Fact]
        public void TestFunctionTakesPrimitiveGenericParamsThrows()
        {
            nint a = 1;
            double b = 2.3;
            Assert.Throws<SwiftRuntimeException>(() => GenericTests.AcceptsGenericParametersAndThrows(a, b));
        }

        [Fact]
        public void TestFunctionTakesPrimitiveGenericParams()
        {
            nint a = 1;
            double b = 2.3;
            var result = GenericTests.AcceptsGenericParameters(a, b);
            Assert.Equal(0, result);
        }

        [Fact]
        public void TestFunctionTakesStructGenericParams()
        {
            var a = new FrozenStruct(1, 2);
            var b = new NonFrozenStruct(3, 4);
            var result = GenericTests.AcceptsGenericParameters(a, b);
            Assert.Equal(0, result);
        }

        [Fact]
        public void TestFunctionTakesGenericPrimitiveAndReturnsOne()
        {
            nint a = 1;
            var result = GenericTests.AcceptsGenericParameterAndReturnsGeneric(a);
            Assert.Equal(a, result);

            double b = 2.3;
            var result2 = GenericTests.AcceptsGenericParameterAndReturnsGeneric(b);
            Assert.Equal(b, result2);

            float c = 3.4f;
            var result3 = GenericTests.AcceptsGenericParameterAndReturnsGeneric(c);
            Assert.Equal(c, result3);
        }

        [Fact]
        public void TestFunctionTakesGenericStructAndReturnsOne()
        {
            var a = new FrozenStruct(1, 2);
            var result = GenericTests.AcceptsGenericParameterAndReturnsGeneric(a);
            Assert.Equal(a, result);

            var b = new NonFrozenStruct(3, 4);
            var result2 = GenericTests.AcceptsGenericParameterAndReturnsGeneric(b);
            // No deep comparison for non-frozen structs yet
        }

        [Fact]
        public void TestFunctionTakesMultipleGenericParametersOfSameType()
        {
            var a = new FrozenStruct(1, 2);
            var b = new FrozenStruct(3, 4);
            var result = GenericTests.AcceptsTwoValuesOfTheSameGenericType(a, b);
            Assert.Equal(a, result);
        }

        [Fact]
        public void TestFunctionTakesGenericParameterConstrainedToProtocol()
        {
            var a = new SummableStruct(2, 40);
            var result = GenericTests.AcceptsSummable(a);
            Assert.Equal(42, result);
        }

        [Fact]
        public void TestFunctionTakesMultipleGenericParametersOfSameTypeConstrainedToProtocol()
        {
            var a = new SummableStruct(2, 40);
            var b = new SummableStruct(3, 39);
            var result = GenericTests.AcceptsMultipleGenericParamsOfTheSameTypeConstrainedByProtocol(a, b);
            Assert.Equal(42 + 42, result);
        }

        [Fact]
        public void TestFunctionTakesGenericParameterConstrainedToMultipleProtocols()
        {
            var a = new StructWithMultipleProtocols(43, 177);
            var result = GenericTests.AcceptsMultipleProtocols(a);
            Assert.Equal(43 + 177 + 43 - 177 + 43 * 177, result);
        }

        [Fact]
        public void TestFunctionTakesMultipleGenericParametersConstrainedToMultipleProtocols()
        {
            var a = new StructWithMultipleProtocols(43, 177);
            var b = new StructWithMultipleProtocols(531, 133);
            var result = GenericTests.AcceptsMultipleGenericParamsWithProtocols(a, b);
            Assert.Equal((43 + 177) + (43 * 177) + (531 - 133) + (531 / 133), result);
        }

        [Fact]
        public void TestFunctionTakesMultipleGenericParametersOfDifferentTypesConstrainedByTheSameProtocol()
        {
            var a = new SummableStruct(2, 40);
            var b = new AnotherSummableStruct(3, 39);
            var result = GenericTests.AcceptsMultipleGenericParamsOfDifferentTypesConstrainedByTheSameProtocol(a, b);
            Assert.Equal(42 + 42, result);
        }

        [Fact]
        public void TestFunctionWithGenericParamConstrainedToPAT()
        {
            var a = new IntContainer1(42);
            var result = GenericTests.AcceptsIntContainer(a);
            Assert.Equal(42 * 2, result);

            var b = new IntContainer2(42);
            var result2 = GenericTests.AcceptsIntContainer(b);
            Assert.Equal(42 * 4, result2);
        }
    }
}

File: src/Swift.Bindings/tests/IntegrationTests/FunctionalTests/Generics/GenericTests.swift
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

import Foundation

@frozen
public struct FrozenStruct
{
    public var x: Int
    public var y: Int

    public init(x: Int, y: Int) {
        self.x = x
        self.y = y
    }
}

public struct NonFrozenStruct
{
    public var x: Int
    public var y: Int

    public init(x: Int, y: Int) {
        self.x = x
        self.y = y
    }
}

public func AcceptsGenericParametersAndThrows<T, U> (a: T, b: U) throws -> Int {
    throw NSError()
}

public func AcceptsGenericParameters<T, U> (a: T, b: U) throws -> Int {
    if a is Int && b is Double {
        return 0;
    }

    if a is FrozenStruct && b is NonFrozenStruct {
        return 0;
    }

    throw NSError()
}

public func AcceptsGenericParameterAndReturnsGeneric<T>(a: T) -> T {
    return a;
}

public func AcceptsTwoValuesOfTheSameGenericType<T>(a: T, b: T) -> T {
    return a;
}

public protocol Summable {
    func sum() -> Int
}

public protocol Subtractable {
    func subtract() -> Int
}

public protocol Multiplicable {
    func multiply() -> Int
}

public protocol Dividable {
    func divide() -> Int
}

@frozen
public struct SummableStruct: Summable {
    public var x: Int
    public var y: Int

    public init(x: Int, y: Int) {
        self.x = x
        self.y = y
    }

    public func sum() -> Int {
        return x + y
    }
}

@frozen 
public struct AnotherSummableStruct: Summable {
    public var x: Int
    public var y: Int

    public init(x: Int, y: Int) {
        self.x = x
        self.y = y
    }

    public func sum() -> Int {
        return x + y
    }
}

public func AcceptsSummable<T: Summable>(a: T) -> Int {
    return a.sum()
}

public func AcceptsMultipleGenericParamsOfTheSameTypeConstrainedByProtocol<T: Summable>(a: T, b: T) -> Int {
    return a.sum() + b.sum()
}

public func AcceptsMultipleGenericParamsOfDifferentTypesConstrainedByTheSameProtocol<T: Summable, U: Summable>(a: T, b: U) -> Int {
    return a.sum() + b.sum()
}

public struct StructWithMultipleProtocols: Summable, Subtractable, Multiplicable, Dividable {
    public var x: Int
    public var y: Int

    public init(x: Int, y: Int) {
        self.x = x
        self.y = y
    }

    public func sum() -> Int {
        return x + y
    }

    public func subtract() -> Int {
        return x - y
    }

    public func multiply() -> Int {
        return x * y
    }

    public func divide() -> Int {
        return x / y
    }
}

public func AcceptsMultipleProtocols<T: Summable & Subtractable & Multiplicable>(a: T) -> Int {
    return a.sum() + a.subtract() + a.multiply()
}

public func AcceptsMultipleGenericParamsWithProtocols<T: Summable & Multiplicable, U: Subtractable & Dividable>(a: T, b: U) -> Int {
    return a.sum() + a.multiply() + b.subtract() + b.divide()
}

public protocol Container {
    associatedtype Element
    func increase() -> Element
}

public struct IntContainer1: Container {
    public var value: Int

    public init(value: Int) {
        self.value = value
    }

    public func increase() -> Int {
        return value * 2
    }
}

public struct IntContainer2: Container {
    public var value: Int

    public init(value: Int) {
        self.value = value
    }

    public func increase() -> Int {
        return value * 4
    }
}

public func AcceptsIntContainer<T: Container>(a: T) -> Int where T.Element == Int {
    return a.increase()
}
